knative-operator/pkg/common/service_monitor.go

```diff
@@ -122,6 +123,93 @@ func SetupSourceServiceMonitor(client client.Client, instance *appsv1.Deployment
 	return nil
 }
 
+func SetupServingControlPlaneServiceMonitors(api client.Client, instance *v1alpha1.KnativeServing) error {
+	serviceNames := []string{"activator", "autoscaler", "autoscaler-hpa", "controller", "domain-mapping", "domainmapping-webhook", "webhook"}
+	for _, sName := range serviceNames {
+		// Create service for service monitor
+		serv := &v1.Service{}
+		serv.Name = sName + "-sm-service"
+		serv.Namespace = "knative-serving"
+		serv.Labels = map[string]string{"name": sName + "-sm-service"}
+		serv.OwnerReferences = []metav1.OwnerReference{*metav1.NewControllerRef(instance, instance.GroupVersionKind())}
+		serv.Spec = v1.ServiceSpec{
+			Ports: []v1.ServicePort{
+				{
+					Name:       "https",
+					Port:       int32(8444),
+					TargetPort: intstr.FromInt(8444),
+				}},
+			Selector: map[string]string{"app": sName},
+		}
+		err := api.Create(context.Background(), serv)
+		if err != nil && !apierrs.IsAlreadyExists(err) {
+			return fmt.Errorf("failed to create service monitor service: %w", err)
+		}
+		// Create service monitor
+		var sm = &monitoringv1.ServiceMonitor{}
+		sm.Name = sName + "-sm"
+		sm.Namespace = "knative-serving"
+		sm.OwnerReferences = []metav1.OwnerReference{*metav1.NewControllerRef(instance, instance.GroupVersionKind())}
+		sm.Spec = monitoringv1.ServiceMonitorSpec{
+			Endpoints: []monitoringv1.Endpoint{
+				{
+					Port:            "https",
+					Scheme:          "https",
+					BearerTokenFile: "/var/run/secrets/kubernetes.io/serviceaccount/token",
+					TLSConfig: &monitoringv1.TLSConfig{
+						SafeTLSConfig: monitoringv1.SafeTLSConfig{
+							InsecureSkipVerify: true,
+						}},
+				}},
+			NamespaceSelector: monitoringv1.NamespaceSelector{
+				MatchNames: []string{sm.Namespace},
+			},
+			Selector: metav1.LabelSelector{
+				MatchLabels: map[string]string{"name": sName + "-sm-service"},
+			}}
+
+		if err := api.Create(context.Background(), sm); err != nil {
+			if err == metrics.ErrServiceMonitorNotPresent {
```

Comment:
I don't think this error will be returned :thinking:. That error might be returned by some osdk helpers, but not by the generic API. You should get a more generic ResourceDoesntExist error here.

Reply:
Yes I will update this.